### DISCUSSION TITLE
Sensor Management

### DISCUSSION THREAD

**Discussion by jmbarr** [dstl/Stone-Soup#441] [Ideas]
Related to #326 and #435 we (@nperree-dstl, @oharrald-Dstl, @jjosborne-dstl and @sdhiscocks) had a recent sprint in order to refine our understanding of sensor management issues and define some useful classes for Stone Soup. Some insights derived (by me) follow. I expect others will add or correct me as necessary.

1. A Sensor Manager is minimally built from a set of sensors which it will manage$^a$ and a `choose_action()` function$^b$. 
2. Sensors should be capable of returning 'Actions' realisable within a time window. Sensors should not be required to be aware of target/scenario details and other sensors. The job of deciding the correct combination of sensor actions is for the sensor manager.
3.  `choose_action()` may use a `reward_function()` which is passed as a parameter to the sensor manager to test actions from subsets$^c$ of its sensors and return a collection of best, or maybe just good, actions to its sensors
4. An `Action` is an object interpretable by a sensor which ensures that the sensor changes its internal state at (or before) a particular time (in anticipation of that sensor making an observation)$^d$.

a. A sensor manager might also be configured to manage other sensor managers. Longer-term perhaps we could create a parent class for both sensor and sensormanager classes. We haven't tried this, nor considered it even.
b. the reward function is a `Callable` type (so can be a function, a class, or a class method)
c. If we are to provide sensor managers that can be used across different scenarios (as is our goal) then `reward_function()` will have to have a flexible enough interface to cope with many scenarios. For example a brute-force sensor management algorithm should execute the same `choose_actions()` (cycle through all possible actions - select the best) for scenarios as diverse as tracking, classification, search, etc, with concomitantly variable reward calculations
d. The sequencing here is difficult and depends whether you're running a simulation or dealing with real life. Current thinking is that we need a sensor `add_actions(Action, timestamp)` function which places an Action onto a sensors to-do list together with a time when it should be undertaken. It's then for the simulator/emulator/sensor itself to ensure that the measurement is executed at that time. 

Next steps:
1. Figure out how to put the above into a diagram (I think that would help)
2. Modify a minimal number of sensors to return/receive actions
3. Finalise the SensorManager base class
4. Code up and introduce a RandomSensorManager and BruteForceSensorManager classes into the code. (Note that both will be useless long-term but will hopefully tease out the difficulties in using sensor managers in different scenarios)
5. Create notebook(s) that shows how to use 4 above

Happy to receive thoughts and contributions

**Comment by sdhiscocks** [comment]
> 2\. Sensors should be capable of returning 'Actions' realisable within a time window. Sensors should not be required to be aware of target/scenario details and other sensors. The job of deciding the correct combination of sensor actions is for the sensor manager.

I think it will be useful for `Actions` / action generators (name we used in sprint to represent possible bounds to select a single action), would have a method (where applicable) to return the action which results in estimated state being possibly observed. e.g. `action_generator.action_to_observe(state: State) -> Optional[Action]`. This way sensor manager doesn't need to work out angles, etc. I imagine this will be simply a case of applying measurement model (state space to measurement space), checking value is in range of options, and returning action e.g. That sate would be at angle _φ_, the sensor can change my dwell centre to that angle, and this is the action required: rotate dwell centre _x_ degrees in _n_ seconds.
